client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClient.kt

```
@@ -143,26 +143,32 @@ class RPCClient<I : RPCOps>(
 
-            val serverProtocolVersion = ops.protocolVersion
-            if (serverProtocolVersion < rpcConfiguration.minimumServerProtocolVersion) {
-                throw RPCException("Requested minimum protocol version (${rpcConfiguration.minimumServerProtocolVersion}) is higher" +
-                        " than the server's supported protocol version ($serverProtocolVersion)")
-            }
-            proxyHandler.setServerProtocolVersion(serverProtocolVersion)
+                val serverProtocolVersion = ops.protocolVersion
+                if (serverProtocolVersion < rpcConfiguration.minimumServerProtocolVersion) {
+                    throw RPCException("Requested minimum protocol version (${rpcConfiguration.minimumServerProtocolVersion}) is higher" +
+                            " than the server's supported protocol version ($serverProtocolVersion)")
+                }
+                proxyHandler.setServerProtocolVersion(serverProtocolVersion)
 
-            log.debug("RPC connected, returning proxy")
-            object : RPCConnection<I> {
-                override val proxy = ops
-                override val serverProtocolVersion = serverProtocolVersion
-                override fun close() {
-                    proxyHandler.close()
-                    serverLocator.close()
+                log.debug("RPC connected, returning proxy")
+                object : RPCConnection<I> {
+                    override val proxy = ops
+                    override val serverProtocolVersion = serverProtocolVersion
+                    override fun close() {
+                        proxyHandler.close()
+                        serverLocator.close()
+                    }
                 }
+            } catch (exception: Throwable) {
+                proxyHandler.close()
+                serverLocator.close()
+                throw exception
```

That would ignore exceptions, we'd need a way to chain those closures and collect the exceptions that happen when running them. I really don't think we should do this, we should cater for codepaths that do happen under normal operation (for example failing to connect is a valid codepath), other exceptions should be thrown, they indicate bugs. Being so defensive just makes the code unreadable

Also, I made `close` idempotent, added tests for this, and for leaking. That issue refers to the old code, can you test it with this branch please?